File: VOSH/2020-2021/Second_Stage/5.cpp
#include <iostream>
#include <cmath>
#include <map>
#include <algorithm>
#include <vector>
#define M_PI 3.14159265358979323846

using namespace std;

int main()
{
    int n;
    cin >> n;

    int result = 0;
    vector<pair<int, int>> points;
    for (int i = 0; i < n; i++)
    {
        pair<int, int> temp;
        cin >> temp.first >> temp.second;
        points.push_back(temp);
    }

    vector<int> out;

    for (int i = 0; i < n; i++)
    {
        vector<float> angles;
        for (int j = 0; j < n; j++)
        {
            if (i == j)
                continue;
            pair<int, int> p0, p1;
            if (points[i].second > points[j].second)
            {
                p0 = make_pair(points[j].first, points[j].second);
                p1 = make_pair(points[i].first, points[i].second);
            }
            else
            {
                p1 = make_pair(points[j].first, points[j].second);
                p0 = make_pair(points[i].first, points[i].second);
            }

            float atg = atan2((float)p1.second - (float)p0.second, (float)p1.first - (float)p0.first);

            if (atg == M_PI)
                atg = 0.0;
            angles.push_back(atg);
        }
        map<float, int> counters;
        for (auto i : angles)
            ++counters[i];
        int currentMax = 0;
        int arg_max = 0;
        for (auto it = counters.cbegin(); it != counters.cend(); ++it)
        {
            if (it->second > currentMax)
            {
                arg_max = it->first;
                currentMax = it->second;
            }
        }
        out.push_back(currentMax + 1);
    }
    cout << *max_element(out.begin(), out.end()) << endl;
}

File: Grade_10/VOSH/2020-2021/Second_Stage/5.py
import math

n = int(input())
points = [list(map(int, input().split())) for _ in range(n)]

out = []
for i in range(len(points)):
    angles = []
    for j in range(len(points)):
        if i == j: continue

        if points[i][1] > points[j][1]:
            p0 = (points[j][0], points[j][1])
            p1 = (points[i][0], points[i][1])
        else:
            p1 = (points[j][0], points[j][1])
            p0 = (points[i][0], points[i][1])
        atg = math.atan2(p1[1] - p0[1], p1[0] - p0[0])
        if atg == math.pi:
            atg = 0.0
        angles.append(atg)
    out.append(
        angles.count(
            max(
                set(angles),
                key=angles.count
            )
        ) + 1
    )

print(max(out))
